scripts/create_deprecated_module.lean

```diff
@@ -0,0 +1,175 @@
+  let mut msgs := #[]
+  let getHash (n : Nat) := do
+    let log ← IO.Process.run {cmd := "git", args := #["log", "--pretty=oneline", s!"-{n}"]}
+    let some last := log.trim.splitOn "\n" |>.getLast? | throwError "Found no commits!"
+    let commitHash := last.takeWhile (!·.isWhitespace)
+    return commitHash
+  let getFilesAtHash (hash : String) := do
+    let files ← IO.Process.run
+      {cmd := "git", args := #["ls-tree", "-r", "--name-only", hash, "Mathlib/"]}
+    let h : Std.HashSet String := .ofList <| files.splitOn "\n"
+    return h
+  let currentHash ← getHash 1
+  let currentFiles ← getFilesAtHash currentHash
+  msgs := msgs.push m!"{currentFiles.size} files at the current hash {currentHash}\n"
+  let pastHash ← getHash n
+  let pastFiles ← getFilesAtHash pastHash
+  msgs := msgs.push m!"{pastFiles.size} files at the past hash {pastHash}\n"
+  let onlyPastFiles := pastFiles.filter fun fil ↦
+    fil.takeRight ".lean".length == ".lean" &&
+    !currentFiles.contains fil
+  let noFiles := onlyPastFiles.size
+  msgs := msgs.push
+    m!"{noFiles} Lean file{if noFiles == 1 then "" else "s"} in 'Mathlib' that no longer exist."
+  let deprecation← if let some cmt := comment then mkDeprecation cmt.getString else mkDeprecation
+  msgs := msgs.push ""
+  for fname in onlyPastFiles do
+    let file ← IO.Process.run {cmd := "git", args := #["show", s!"{pastHash}:{fname}"]}
+    let fileHeader ← getHeader fname file false
+    let deprecatedFile := s!"{fileHeader.trimRight}\n\n{deprecation.pretty.trimRight}\n"
+    msgs := msgs.push <| .trace {cls := `Deprecation} m!"{fname}" #[m!"\n{deprecatedFile}"]
+    if write?.isSome then
+      IO.FS.writeFile fname deprecatedFile
+  if write?.isNone && noFiles != 0 then
+    -- We strip trailing comments from `nc` and `comment` to avoid them showing up in the
+    -- regenerated syntax.
+    let nc := nc.map (⟨·.raw.unsetTrailing⟩)
+    let comment := comment.map (⟨·.raw.unsetTrailing⟩)
+    let stx ← `(command|#create_deprecated_modules $[$nc:num]? $[$comment:str]? write)
+    msgs := msgs.push
+      m!"The files were not deprecated. Use '{stx}' \
+        if you wish to deprecate them."
+  logInfoAt tk <| .joinSep msgs.toList "\n"
+
+/-
+Uncomment the `#create_deprecated_modules` line below to get started.
+* `10` should likely be some small number, though its exact value may not be too important;
+* omitting `"a comment here"` is equivalent to using `"Auto-generated deprecation"`
+  while using the empty string `""` eliminates the comment entirely;
+* uncomment `write` only when you are satisfied that the deprecations look correct!
+-/
+--#create_deprecated_modules 10 "a comment here" --write
+
+/--
+info: /-
+Copyright (c) 2025 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+--import Mathlib.Init
+import Lean.Elab.Command
+import Mathlib.Tactic.Linter.DeprecatedModule
+-/
+#guard_msgs in
+run_cmd
```

Comment:
I wonder if there's a way to test the actual deprecation output; that seems harder to arrange, though. I'm fine with just these tests!

Reply:
I could extract lots of small functions that perform each part of the deprecation and test them separately, but I am not sure that it would improve the script.

Right now, the script is pretty verbose in saying what it is doing and what the outcome is, so testing "live" should be straightforward.